contracts/VeJoeStaking.sol

```diff
@@ -0,0 +1,267 @@
+    /// receiving boosted benefits, in parts per 100.
+    /// @dev Specifically, user has to deposit at least `boostedThreshold/100 * totalStakedJoe` JOE.
+    /// The only exception is the user will also receive boosted benefits if its their first
+    /// time staking.
+    uint256 public boostedThreshold;
+
+    /// @notice The length of time a user receives boosted benefits
+    uint256 public boostedDuration;
+
+    mapping(address => UserInfo) public userInfos;
+
+    event Deposit(address indexed user, uint256 amount);
+    event Withdraw(address indexed user, uint256 amount);
+    event Claim(address indexed user, uint256 amount);
+
+    /// @notice Initialize with needed parameters
+    /// @param _joe Address of the JOE token contract
+    /// @param _veJoe Address of the veJOE token contract
+    /// @param _baseGenerationRate Rate of veJOE generated per sec per JOE staked
+    /// @param _boostedGenerationRate Boosted rate of veJOE generated per sec per JOE staked
+    /// @param _boostedThreshold Percentage of total staked JOE user has to deposit to be boosted
+    /// @param _boostedDuration Length of time a user receives boosted benefits
+    function initialize(
+        IERC20Upgradeable _joe,
+        VeJoeToken _veJoe,
+        uint256 _baseGenerationRate,
+        uint256 _boostedGenerationRate,
+        uint256 _boostedThreshold,
+        uint256 _boostedDuration
+    ) public initializer {
+        require(address(_joe) != address(0), "VeJoeStaking: unexpected zero address for _joe");
+        require(address(_veJoe) != address(0), "VeJoeStaking: unexpected zero address for _veJoe");
+        require(
+            _boostedGenerationRate > _baseGenerationRate,
+            "VeJoeStaking: expected _boostedGenerationRate to be greater than _baseGenerationRate"
+        );
+        require(_boostedThreshold <= 100, "VeJoeStaking: expected _boostedThreshold to be less than or equal to 100");
+
+        __Ownable_init();
+
+        maxCap = 100;
+        joe = _joe;
+        veJoe = _veJoe;
+        baseGenerationRate = _baseGenerationRate;
+        boostedGenerationRate = _boostedGenerationRate;
+        boostedThreshold = _boostedThreshold;
+        boostedDuration = _boostedDuration;
+    }
+
+    /// @notice Set maxCap
+    /// @param _maxCap the new maxCap
+    function setMaxCap(uint256 _maxCap) external onlyOwner {
+        require(_maxCap > 0, "VeJoeStaking: expected new _maxCap to be greater than 0");
+        maxCap = _maxCap;
+    }
+
+    /// @notice Set baseGenerationRate
+    /// @param _baseGenerationRate the new baseGenerationRate
+    function setBaseGenerationRate(uint256 _baseGenerationRate) external onlyOwner {
+        require(
+            _baseGenerationRate < boostedGenerationRate,
+            "VeJoeStaking: expected new _baseGenerationRate to be less than boostedGenerationRate"
+        );
+        baseGenerationRate = _baseGenerationRate;
```

Comment:
It would be a "nice to have" if we had events that got emitted when important variables are updated.